### PR TITLE
[diffusion] fix: stop silently dropping weights in update_weights_fro…

### PR DESCRIPTION
**Pull request by firefly-silvers**
…m_disk

update_weights_from_disk reported success while discarding weights:

  - VAE BatchNorm buffers were never loaded (named_parameters() excludes buffers), leaving stale latent-normalisation statistics that scale every decode
  - text encoders were routed through the generic name-mapping path, which cannot strip their "model." prefix or fuse q/k/v, so no weight loaded
  - the API counted a module as updated if nothing raised, so all of this was invisible

Changes:
  - include named_buffers() alongside named_parameters()
  - route TextEncoder modules to their own load_weights(), materialising layerwise-offloaded encoders first (same mechanism the LoRA IPC path already uses)
  - report per-module loaded counts in the response, and warn on zero

Known limitation: FSDP-sharded text encoders are still skipped -- load_weights runs before shard_model() at startup so it has no DTensor branch. Marked with a TODO; now visible as "text_encoder: 0" instead of silent.

Verified on 1x NVIDIA L4 with FLUX.2-klein-base-4B: silently dropped weights across the 5 tests in test_update_weights_from_disk.py went 2409 -> 398, all tests passing. Two CPU regression tests added, each verified to fail when its fix is reverted.

Partially addresses #31924.

## Motivation

`update_weights_from_disk` returned `success: true` while silently discarding weights. On a single-GPU FLUX.2-klein-base-4B run, the 5 tests in `test_update_weights_from_disk.py` dropped **2409 tensors** and every request still reported success.

This matters because the API exists for RL refit: a training loop can swap in new weights, get a 200 back, and keep serving the old ones. Three independent causes:
1. **VAE buffers were never loaded.** The reload path built its lookup from `named_parameters()`, which excludes buffers. For the FLUX.2 VAE the BatchNorm is `affine=False` (`autoencoder_kl_flux2.py:109`), so `running_mean` / `running_var` are the layer's *entire* state — and they set the scale and shift applied to every decode (`configs/pipeline_configs/flux.py:703-717`). A reloaded VAE kept stale normalisation statistics.

2. **Text encoders loaded nothing at all.** They were routed through the generic name-mapping path, which cannot strip their `model.` prefix or apply `stacked_params_mapping` fusion. Startup does both inside the encoder's own `load_weights` (`encoders/qwen3.py:431,467`); the reload path never called it, so all 398 tensors missed.

3. **Nothing counted what landed.** `_apply_weights` marked a module updated if no exception was raised, so 1 and 2 were invisible. Startup already rejects this — it compares `weights_to_load - loaded_weights` and refuses to continue with uninitialised weights (`text_encoder_loader.py:460-467`). Reload never got the same treatment.

Partially addresses #31924. See the note at the bottom on that issue's headline symptom.

## Modifications

- **Include buffers.** `_load_weights_into_module` now builds its lookup from `named_parameters()` **and** `named_buffers()`.

- **Dispatch text encoders to their own loader.** Three-way branch, offload checked first (an offloaded text encoder satisfies more than one condition). Layerwise-offloaded encoders are materialised via `disable_offload()` / `enable_offload()` first, since `load_weights` cannot write into `torch.empty((1,))` placeholders. That is the same mechanism the LoRA IPC path already uses (`weights_updater.py:631` → `lora_pipeline.py:208-219`), applied to a ~1 GB encoder rather than the 7.8 GB DiT.

- **Guard the materialisation on free VRAM.** Materialising is not free: measured at **7.32 GB** for the FLUX.2-klein text encoder, against **8.27 GB** free at that point on a 24 GB L4 (the DiT and VAE are already resident). Without a guard this could OOM on a tighter card or a larger encoder. `_can_materialize_module` compares the module's offloaded byte total against available memory and, if it will not fit, **falls back to the existing offload path** — the module is skipped exactly as it is today, and reports `0` so the skip stays visible. Never worse than current behaviour.

- **`LayerwiseOffloadManager.resident_bytes` property** (`layerwise_offload.py`). Sums both CPU storages — the consolidated per-dtype buffers and the strided tensors used for layout-sensitive (quantized) weights. No behaviour change; added so the size calculation lives with the storage it describes rather than reaching into the manager's internals from `weights_updater`. Summing over `named_parameters()` would report almost nothing, since those are `(1,)` placeholders under offload.

- **Report what landed.** `_load_weights_into_module` and `load_weights_into_model` return the names they applied; the response message now carries per-module counts, and a zero count logs a warning:
  ```
  Updated 3 modules (text_encoder: 290, transformer: 169, vae: 251).
  ```
**Blast radius.** The transformer path is unchanged — it is the one path that already worked. `update_weights_from_tensor` (the RL tensor-IPC entry point, which defaults to `transformer` only) is untouched. The two modules this changes were 100% broken before, so there is nothing to regress.

**On the guard's constants.** `_MATERIALIZE_VRAM_MARGIN = 1.2` and a 1 GB floor are heuristics. They are grounded in one measurement: `resident_bytes` reports 6.77 GB while the observed VRAM delta was 7.32 GB — an ~8% underestimate, presumably allocator block rounding — so a margin is needed, and 1.2× covers it with ~0.8 GB to spare. Happy to take different values, or a different signal entirely, if there is an established convention here I have missed.

**Known limitation.** FSDP-sharded text encoders are still skipped. `load_weights` runs before `shard_model()` at startup, so it has no DTensor branch and raises `aten.copy_.default got mixed torch.Tensor and DTensor` if called on one. Those modules are excluded by an explicit guard and fall back to the generic path. Marked with a `TODO` and left for a follow-up - closing it needs either DTensor handling inside each model's `load_weights`, or the `fsdp_load.py` pattern (weight loader into a plain temp tensor, then `distribute_tensor`), and no `unshard`/`reshard` precedent exists in `multimodal_gen` today. Thanks to the per-module counts it is no longer silent — it surfaces as `text_encoder: 0`.

## Accuracy Tests

`test_update_weights_from_disk.py`, 1×NVIDIA L4, `FLUX.2-klein-base-4B`, all 5 tests passing before and after (the tests never caught this — they only assert the transformer checksum).

Silently dropped weights across the suite:

| module | before | after |
|---|---|---|
| `Qwen3ForCausalLM` (text encoder) | 1990 | **0** |
| `AutoencoderKLFlux2` (vae) | 21 | **0** |
| `Flux2Transformer2DModel` | 0 | 0 |
| `FSDPQwen3ForCausalLM` | 398 | 398 (known limitation) |
| **total** | **2409** | **398** |

Per-module counts now reported, cross-checked against the checkpoints: transformer 169 = the checkpoint's 169 tensors; vae 251 = 248 parameters + 3 BatchNorm buffers; text_encoder 290 = its parameter count after q/k/v and gate/up fusion.

Two CPU unit tests added (`test/unit/test_weights_updater.py`), exercising `update_weights_from_disk` end to end against fake modules. **Each was verified to fail when its own fix is reverted, and only its own** — reverting the buffer union fails the buffer test alone (`vae: 2`, not 5); reverting the text-encoder branch fails that test alone (`text_encoder: 0`, not 2), which is the same signature as the production bug.

Verified at single-GPU only. Tensor-parallel and multi-GPU FSDP sharding of the text encoder are not exercised by this setup.

## Speed Tests and Profiling

No inference-path change; the modified code runs only during `update_weights_from_disk`, which is serialised by the scheduler and never concurrent with generation.

Two reload-path costs, both bounded:

1. **Materialising an offloaded text encoder** temporarily makes its weights GPU-resident - measured at **7.32 GB** for FLUX.2-klein — for the duration of the update, then re-offloads in a `finally`. This is now guarded on free VRAM (see Modifications); if it will not fit, the update falls back rather than risking an OOM. For reference, the existing LoRA IPC path performs the same materialisation on the 7.8 GB DiT without a guard.

2. **The free-memory query calls `torch.cuda.empty_cache()`** (the default for `current_platform.get_available_gpu_memory`). That releases cached blocks back to the driver, so the subsequent materialisation allocates through `cudaMalloc` rather than reusing the caching allocator. It runs once per offloaded-text-encoder update - roughly 6–7 times across the test suite. I kept it because passing `empty_cache=False` underreports free memory (cached blocks look occupied), and the guard already approves with only ~0.15 GB of slack on this setup; a pessimistic reading would make it reject when it should not. Open to the opposite trade-off if reviewers prefer it.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations). 
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

---

**On #31924's headline symptom.** That issue reports a shape-mismatch 400 on the *transformer*. I could not reproduce it: all 5 tests pass on NVIDIA L4 at commit `b61cb5f9de`, on a `weights_updater.py` unchanged since #31263 — i.e. the same code that failed on AMD. The transformer had zero skips, because that checkpoint's 169 tensors already use sglang's parameter names, so `param_names_mapping` never runs. Details posted to the issue. This PR therefore fixes the secondary defect and does **not** unblock re-enabling the ROCm skip from #31925.

---
### CI States

Latest PR Test (Base): :x: [Run #31022203118](https://github.com/sgl-project/sglang/actions/runs/31022203118)
Latest PR Test (Extra): :x: [Run #31022203914](https://github.com/sgl-project/sglang/actions/runs/31022203914)

### CODE REVIEW

**Comment by firefly-silvers** [comment]
@HaiShaw @mickqian Can I get some reviews?
